raft-tech/TANF-app
# Local Alert Manager

## Summary of Changes
- Added MVP implementation of Alertmanager
- Integrated Alertmanager with SendGrid for email alerts
- Added initial prometheus alerting rules based on available metrics
- Added some stub work for when this is deployed
- Updated the `Logs` dashboard to also display uptime/availability
- Updated frontend to allow proxy pass to Alertmanger UI
Pull request closes #3242 

## How to Test
- Before you `up` everything. You need to add your email (instead of mine) and a SENDGRID_API_KEY to alertmanager.yml.
  - On line 6 in `alertmanager.yml` replace `{{ sendgrid_api_key }}` with a valid api key.
  - On line 104 in `alertmanager.yml` replace my emails with your email(s).


```
cd tdrs-backend && docker-compose up --build
```
1. Open the `alertmanager` container logs and verify no new alerts are firing (note you could see a message like: `level=debug component=dispatcher msg="Received alert" alert="Local Backend Down[7649f89][active]"` initially and that it is immediately resolved after: `level=debug component=dispatcher msg="Received alert" alert="Local Backend Down[7649f89][resolved]"`. This happens because of some tight timing tolerances. You will not get an email for this which is expected.
2. Let everything run for a minute or two and verify `alertmanager` is NOT firing any new alerts.
3. Kill  `postgres` or `web` or both containers and watch `alertmanager` start firing alerts. After the alert(s) have fired for at least 1 minute you will receive emails for the alerts. You will not receive another email for the alert for another 5 minutes
4. Restart the container(s) you killed and verify `alertmanager` marks the firing alert(s) as resolved.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [ ] Prometheus connects to Alertmanager
- [ ] Prometheus sends alerts to Alertmanager
- [ ] Alertmanager integrated with SendGrid
- [ ] Alert emails are received from Alertmanager/SendGrid
- [ ] Documentation updated indicating Alertmanager integration

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)


### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)



### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)


### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)


### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)


### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
